main/musl/patches/0001-plumb-in-support-for-externally-provided-allocator-l.patch

```diff
@@ -0,0 +1,292 @@
+* __malloc_atfork
+* __malloc_donate
+* __malloc_tls_teardown
+* __malloc_tls_default
+* __libc_calloc
+* __libc_free
+* __libc_malloc_impl
+* __libc_realloc
+* aligned_alloc
+* malloc_usable_size
+
+It can access the following variables via extern:
+
+* __aligned_alloc_replaced
+* __malloc_replaced
+---
+ Makefile                    | 18 ++++++++++++++----
+ ldso/dynlink.c              |  4 ++--
+ src/env/__init_tls.c        |  8 ++++++--
+ src/internal/pthread_impl.h | 15 ++++++++++++++-
+ src/malloc/calloc.c         |  4 ++++
+ src/malloc/external/empty.h |  1 +
+ src/malloc/libc_calloc.c    |  4 ++++
+ src/thread/pthread_create.c |  7 +++++++
+ 8 files changed, 52 insertions(+), 9 deletions(-)
+ create mode 100644 src/malloc/external/empty.h
+
+diff --git a/Makefile b/Makefile
+index e8cc443..27e388f 100644
+--- a/Makefile
++++ b/Makefile
+@@ -23,6 +23,7 @@ BASE_GLOBS = $(addsuffix /*.c,$(SRC_DIRS))
+ ARCH_GLOBS = $(addsuffix /$(ARCH)/*.[csS],$(SRC_DIRS))
+ BASE_SRCS = $(sort $(wildcard $(BASE_GLOBS)))
+ ARCH_SRCS = $(sort $(wildcard $(ARCH_GLOBS)))
++MALLOC_OBJS = $(sort $(wildcard $(srcdir)/src/malloc/$(MALLOC_DIR)/*.o))
+ BASE_OBJS = $(patsubst $(srcdir)/%,%.o,$(basename $(BASE_SRCS)))
+ ARCH_OBJS = $(patsubst $(srcdir)/%,%.o,$(basename $(ARCH_SRCS)))
+ REPLACED_OBJS = $(sort $(subst /$(ARCH)/,/,$(ARCH_OBJS)))
+@@ -32,8 +33,8 @@ LIBC_OBJS = $(filter obj/src/%,$(ALL_OBJS)) $(filter obj/compat/%,$(ALL_OBJS))
+ LDSO_OBJS = $(filter obj/ldso/%,$(ALL_OBJS:%.o=%.lo))
+ CRT_OBJS = $(filter obj/crt/%,$(ALL_OBJS))
+ 
+-AOBJS = $(LIBC_OBJS)
+-LOBJS = $(LIBC_OBJS:.o=.lo)
++AOBJS = $(LIBC_OBJS) $(MALLOC_OBJS)
++LOBJS = $(LIBC_OBJS:.o=.lo) $(MALLOC_OBJS)
+ GENH = obj/include/bits/alltypes.h obj/include/bits/syscall.h
+ GENH_INT = obj/src/internal/version.h
+ IMPH = $(addprefix $(srcdir)/, src/internal/stdio_impl.h src/internal/pthread_impl.h src/internal/locale_impl.h src/internal/libc.h)
+@@ -131,6 +132,11 @@ $(CRT_OBJS): CFLAGS_ALL += -DCRT
+ 
+ $(LOBJS) $(LDSO_OBJS): CFLAGS_ALL += -fPIC
+ 
++ifneq (mallocng,$(MALLOC_DIR))
++obj/src/malloc/calloc.lo: CFLAGS_ALL += -DLIBC_CALLOC_EXTERNAL
++obj/src/malloc/libc_calloc.lo: CFLAGS_ALL += -DLIBC_CALLOC_EXTERNAL
++endif
++
+ CC_CMD = $(CC) $(CFLAGS_ALL) -c -o $@ $<
+ 
+ # Choose invocation of assembler to be used
+@@ -140,6 +146,10 @@ else
+ 	AS_CMD = $(CC_CMD)
+ endif
+ 
++$(EXTRA_OBJ): $(GENH) $(IMPH)
++	$(CC) -I$(srcdir)/mimalloc/include $(CFLAGS_ALL) -std=gnu11 -fPIC -O3 -DNDEBUG -fvisibility=hidden -isystem `$(CC) -print-resource-dir`/include -c -o $(EXTRA_OBJ) $(srcdir)/mimalloc/src/mimalloc.c
++	sh $(srcdir)/mimalloc-verify-syms.sh $(EXTRA_OBJ)
++
+ obj/%.o: $(srcdir)/%.s
+ 	$(AS_CMD)
+ 
+@@ -158,11 +168,11 @@ obj/%.lo: $(srcdir)/%.S
+ obj/%.lo: $(srcdir)/%.c $(GENH) $(IMPH)
+ 	$(CC_CMD)
+ 
+-lib/libc.so: $(LOBJS) $(LDSO_OBJS)
++lib/libc.so: $(LOBJS) $(LDSO_OBJS) $(EXTRA_OBJ)
+ 	$(CC) $(CFLAGS_ALL) $(LDFLAGS_ALL) -nostdlib -shared \
+ 	-Wl,-e,_dlstart -o $@ $(LOBJS) $(LDSO_OBJS) $(LIBCC)
```

Comment:
@q66 shouldn't we have something like 
```
-Wl,-e,_dlstart -o $@ $(LOBJS) $(EXTRA_OBJ) $(LDSO_OBJS) $(LIBCC)
```

here ? You seem to be depending on the lazy and repeated evaluation of `$(LOBJS)`. The `$(LOBJS)` in the rule prerequisite will be different from the `$(LOBJS)` in the rule recipe...